### PR TITLE
op-batcher: Move decision about data availability type to channel submission time

### LINKED ISSUE #11609: op-batcher: Assess da-type choice when submitting a channel
## Context

Since https://github.com/ethereum-optimism/optimism/pull/11219 the batcher can dynamically switch between calldata and blobs, depending on the economically best option _at the time of starting a new channel_. This works reasonably well for high-throughput chains, where the delay between starting and submitting a channel is usually only a few minutes. But it is still not optimal, as the most economical option might have changed since building up the channel. Furthermore, for low-throughput chains, this behavior is useless, as a da-type decision delay of multiple hours is ineffective. And lastly, the decision is currently made under the assumption that the channel will be full to its size limit. But the channel may not be full, e.g. using less frames (=blobs) with a half-empty last frame because of a short max-channel-duration setting that closed it (especially relevant on testnets).

## Change Description

To address these shortcomings, the batcher should perform a second/the check at the time that it would start submitting a channel (i.e., when starting to post the first batcher transaction for that channel (almost always a channel fits into a single batcher transaction, unless the first block added to it doesn't fit into the `target-num-frames` frames)), and possibly rebuild the channel using different da-type settings.

The goal of this improvement is to always select the best da-type for a channel at the time of tx submission, and make the dynamic da-type selection feature useable for low-throughput chains.

## Implementation Hints

It probably makes most sense to _move_ the channel config retrieval via the `ChannelConfigProvider` from the point at which a new channel is created (https://github.com/ethereum-optimism/optimism/blob/a36291dc345382094ba971ffd4cd5987e84b989f/op-batcher/batcher/channel_manager.go#L206) to the point at which the first frame of any channel is submitted (places where we return non-empty `txData` inside [`channelManager.TxData()`](https://github.com/ethereum-optimism/optimism/blob/a36291dc345382094ba971ffd4cd5987e84b989f/op-batcher/batcher/channel_manager.go#L152C10-L152C32)), and then reuse that config for the following channel. This has two advantages. 
1. This adds the check for the correct channel config at the point when submission happens. Most of the time this check will confirm the current configuration.
2. The channel config of the previous channel is the best estimator for the correct channel config of the following channel, when the exact data content isn't known yet.

Note that a special case is a fresh batcher restart, at which point an estimated configuration has to be retrieved once before the first channel. But this could happen in the batcher's initialization path.

### Channel Rebuilding

If the check doesn't confirm the channel config that's currently in use, the channel should be rebuilt using the new config. One good option to accomplish this may be to reinsert the blocks into the L2 blocks queue inside the `channelManager`, so that the existing code path is then used to build the channel with the new channel config. This has the advantage that e.g. if switching from calldata to blobs, the channel would suddenly have more space, so it wouldn't be immediately submitted and just continue to fill up regularly.

### `ChannelConfigProvider`

It's important to note that the current [`ChannelConfigProvider`](https://github.com/ethereum-optimism/optimism/blob/develop/op-batcher/batcher/channel_config_provider.go) has a single method to retrieve the next estimated channel config, and the implementation of the dynamic provider always assumes full channels. This is not optimal for the proposed change, as we will know the exact channel data at the point of submission, so we can make an exact fee calculation. To enable this, I suggest we add a second method to the `ChannelConfigProvider` to make its interface look similar to the following:
```go
type ChannelConfigProvider interface {
	ChannelConfigFull() ChannelConfig // note: renamed old method that estimates full channels, would be used only at startup
	ChannelConfig(data []byte) ChannelConfig // new method, takes exact data
}
```
The new method can take as a parameter the exact data and use `core.IntrinsicGas` to exactly calculate the calldata gas in the dynamic provider, instead of this manual calculation:
https://github.com/ethereum-optimism/optimism/blob/a36291dc345382094ba971ffd4cd5987e84b989f/op-batcher/batcher/channel_config_provider.go#L66-L68

Otherwise, the existing cost comparison code at `DynamicEthChannelConfig.ChannelConfig()` should be as much reused between the two new methods as possible.

### Channel Config Immutability

One important rule that should be observed is to not change the DA-type and channel configuration once the first frame of a channel started being submitted. This is important a) because otherwise we might end up trying to replace a transaction with a different tx type, which is not allowed and would get the batcher into a blocked tx recovery mode and b) in view of the upcoming Holocene features regarding [Strict Batch Ordering](https://github.com/ethereum-optimism/optimism/issues/11318) would invalidate future batcher transactions that might already be in the tx-pool because of parallel transaction sending.

Hardening the batcher's behavior during restarts should be done as a separate task, possibly introducing some form of persistence of in-flight transactions.

### PR DESCRIPTION
**Pull request by geoknee**
Closes #11609 

For now, I decided to not make any changes to the decision logic itself (only to when it is triggered and what happens when a change is necessary). 

Changing the trigger for the decision opens the opportunity to make a more accurate decision about which DA type to choose, since some transaction data is now in scope. However, this would not address the larger assumption in the current logic, which is about how many blobs are typically included in a tx. This can be lower than the target number on chains with low throughput as channels must be closed before they time out. To address this assumption, we could store the previous number of blobs per tx to make a better estimate than the current one (which assumes we always include the target number). This is saved for future work. 

Tests: the modified implementation passes existing tests, including an end to end test for switching DA type.

### CODE REVIEW

File: op-batcher/batcher/channel_config_provider.go [hunk 1]
```diff
@@ -89,7 +93,7 @@ func (dec *DynamicEthChannelConfig) ChannelConfig() ChannelConfig {
 		"blob_data_bytes", blobDataBytes, "blob_cost", blobCost,
 		"cost_ratio", costRatio)
 
-	if ay.Cmp(bx) == 1 {
+	if ay.Cmp(bx) > 0 {
```
Comment (sebastianst) [inline op-batcher/batcher/channel_config_provider.go]: This is the same, why the change?

File: op-batcher/batcher/channel_manager.go [hunk 2]
```diff
@@ -146,10 +150,47 @@ func (s *channelManager) nextTxData(channel *channel) (txData, error) {
...
 func (s *channelManager) TxData(l1Head eth.BlockID) (txData, error) {
+	data, err := s.txData(l1Head)
+	if s.currentChannel == nil {
+		s.log.Trace("no current channel")
+		return data, err
+	}
+	assumedBlobs := s.currentChannel.cfg.UseBlobs
+	newCfg := s.cfgProvider.ChannelConfig()
+	if newCfg.UseBlobs == assumedBlobs {
+		s.log.Info("Recomputing optimal ChannelConfig: no need to switch DA type",
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: This should rather be a debug log IMO

File: op-batcher/batcher/channel_manager.go [hunk 3]
```diff
@@ -146,10 +150,47 @@ func (s *channelManager) nextTxData(channel *channel) (txData, error) {
...
+//
+// It will generate the tx data internally, and decide whether to switch DA type
+// automatically. When switching DA type, the channelManager state will be rebuilt
+// with a new ChannelConfig.
 func (s *channelManager) TxData(l1Head eth.BlockID) (txData, error) {
+	data, err := s.txData(l1Head)
+	if s.currentChannel == nil {
+		s.log.Trace("no current channel")
+		return data, err
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: 1. I'd rather check `err != nil`, as usual. We shouldn't proceed if this returns non-nil.
2. We should check here whether the channel this `data` comes from already has a frame in submission. If yes, early return. Probably easiest is to return the `channel` together with the `txData` by `txData()`.
3. The `currentChannel` isn't necessarily the channel where the `data` comes from.

File: op-batcher/batcher/channel_manager.go [hunk 4]
```diff
@@ -414,3 +460,28 @@ func (s *channelManager) Close() error {
 	}
 	return nil
 }
+
+// Rebuild rebuilds the channel manager state by
+// rewinding blocks back from the channel queue, and setting the defaultCfg.
+func (s *channelManager) Rebuild(newCfg ChannelConfig) error {
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: Maybe we rename this to `Requeue` or similar, because it doesn't build anything, but puts the blocks back from the channels into the queue?

File: op-batcher/batcher/channel_manager.go [hunk 5]
```diff
@@ -414,3 +460,28 @@ func (s *channelManager) Close() error {
...
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.log.Info("Rebuilding channelManager state", "UseBlobs", newCfg.UseBlobs)
+	newChannelQueue := []*channel{}
+	blocksToRequeueInChannelManager := []*types.Block{}
+	for _, channel := range s.channelQueue {
+		if len(channel.pendingTransactions) > 0 {
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: You'd also need to check `channel.confirmedTransactions`. But even better, there's already a helper `channel.NoneSubmitted()`.

File: op-batcher/batcher/channel_manager.go [hunk 6]
```diff
@@ -146,10 +150,47 @@ func (s *channelManager) nextTxData(channel *channel) (txData, error) {
...
+// It will generate the tx data internally, and decide whether to switch DA type
+// automatically. When switching DA type, the channelManager state will be rebuilt
+// with a new ChannelConfig.
 func (s *channelManager) TxData(l1Head eth.BlockID) (txData, error) {
+	data, err := s.txData(l1Head)
+	if s.currentChannel == nil {
+		s.log.Trace("no current channel")
+		return data, err
+	}
+	assumedBlobs := s.currentChannel.cfg.UseBlobs
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: Couldn't we compare to the `s.defaultCfg` here?

**Review by sebastianst — CHANGES_REQUESTED** [review]
We should definitely add a few unit tests of `TxData()` and the requeue function (currently called `Rebuild`).

**Comment by zhiqiangxu** [comment]
If the Data Availability type is determined at submission time, how are `MaxFrameSize` and `MaxFramesPerTx` decided? I believe there's a kind of interdependency here.

**Comment by geoknee** [comment]
> If the Data Availability type is determined at submission time, how are `MaxFrameSize` and `MaxFramesPerTx` decided? I believe there's a kind of interdependency here.

In dynamic DA mode, the batcher maintains two configs (one for each DA type blobs and calldata). One of those configs is the default (i.e. initially blobs, but generally whatever was used when the previous channel was submitted). The channel can then be built optimistically with the default config. When it comes to submit the channel, if the data availability type needs to change, channels which haven't been submitted will be discarded and rebuilt with the appropriate config (including params like `MaxFrameSize`. Does that make sense to you?

**Comment by zhiqiangxu** [comment]
> > If the Data Availability type is determined at submission time, how are `MaxFrameSize` and `MaxFramesPerTx` decided? I believe there's a kind of interdependency here.
> 
> In dynamic DA mode, the batcher maintains two configs (one for each DA type blobs and calldata). One of those configs is the default (i.e. initially blobs, but generally whatever was used when the previous channel was submitted). The channel can then be built optimistically with the default config. When it comes to submit the channel, if the data availability type needs to change, channels which haven't been submitted will be discarded and rebuilt with the appropriate config (including params like `MaxFrameSize`. Does that make sense to you?

Yeah it makes sense, thank you!

**Comment by sebastianst** [comment]
> In dynamic DA mode, the batcher maintains two configs (one for each DA type blobs and calldata). One of those configs is the default (i.e. initially blobs, but generally whatever was used when the previous channel was submitted). The channel can then be built optimistically with the default config. When it comes to submit the channel, if the data availability type needs to change, channels which haven't been submitted will be discarded and rebuilt with the appropriate config (including params like `MaxFrameSize`. Does that make sense to you?

Worth adding that we're effectively still choosing the config at the same point in time when we create the channel. That's because right after a channel is being submitted, a new channel is created, and then we choose the config that was just selected 2 seconds ago.

File: op-batcher/batcher/channel_config_provider.go (hunk 1 above)
Comment (geoknee) [reply op-batcher/batcher/channel_config_provider.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69

File: op-batcher/batcher/channel_manager.go (hunk 2 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69

File: op-batcher/batcher/channel_manager.go (hunk 3 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69 but I ended up refactoring this more comprehensively.

File: op-batcher/batcher/channel_manager.go (hunk 4 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69

File: op-batcher/batcher/channel_manager.go (hunk 5 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69

File: op-batcher/batcher/channel_manager.go (hunk 6 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: e299a7922d7d8ef54710b8c234386ce41fd26b69

File: op-batcher/batcher/channel_manager_test.go [hunk 7]
```diff
@@ -483,3 +484,174 @@ func TestChannelManager_ChannelCreation(t *testing.T) {
...
+// state is equal to the snapshot. It repeats this for a channel
+// which has a pending transaction and verifies that Requeue is then
+// a noop.
+func TestChannelManager_Requeue(t *testing.T) {
+	l := testlog.Logger(t, log.LevelCrit)
+	cfg := channelManagerTestConfig(100, derive.SingularBatchType)
+	m := NewChannelManager(l, metrics.NoopMetrics, cfg, defaultTestRollupConfig)
+
+	// Seed channel manager with blocks
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
```
Comment (geoknee) [inline op-batcher/batcher/channel_manager_test.go]: Probably better to fix this to make the test more deterministic.

File: op-batcher/batcher/channel_manager.go [hunk 8]
```diff
@@ -139,9 +142,13 @@ func (s *channelManager) removePendingChannel(channel *channel) {
...
 func (s *channelManager) nextTxData(channel *channel) (txData, error) {
-	if channel == nil || !channel.HasTxData() {
+	if channel == nil {
+		s.log.Trace("no channel")
+		return txData{}, io.EOF
+	}
+	if !channel.HasTxData() {
 		s.log.Trace("no next tx data")
-		return txData{}, io.EOF // TODO: not enough data error instead
+		return txData{}, ErrInsufficientData
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: There may be a few places where callers still assume that the error is `io.EOF`. I don't see error checks changed to the new error type, except for in tests. I'd prefer if we do this change to a custom error type in a separate PR, or is it necessary for the scope of this PR?

File: op-batcher/batcher/channel_manager.go [hunk 9]
```diff
@@ -139,9 +142,13 @@ func (s *channelManager) removePendingChannel(channel *channel) {
 // nextTxData dequeues frames from the channel and returns them encoded in a transaction.
 // It also updates the internal tx -> channels mapping
 func (s *channelManager) nextTxData(channel *channel) (txData, error) {
-	if channel == nil || !channel.HasTxData() {
+	if channel == nil {
+		s.log.Trace("no channel")
+		return txData{}, io.EOF
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: note that you still return the `io.EOF` error here, not the new error type.

File: op-batcher/batcher/channel_manager.go [hunk 10]
```diff
@@ -154,43 +161,49 @@ func (s *channelManager) nextTxData(channel *channel) (txData, error) {
...
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	channel, err := s.getReadyChannel(l1Head)
 	if err != nil {
 		return emptyTxData, err
 	}
-	if channel != nil && !channel.NoneSubmitted() {
+	if channel == nil {
+		panic("nil channel and nil err returned from getReadyChannel")
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: Why this explicit panic check? the next line will panic anyways if `channel` was `nil`. I also fail to see how `getReadyChannel` could return `nil, nil`.

File: op-batcher/batcher/channel_manager.go [hunk 11]
```diff
@@ -202,11 +215,15 @@ func (s *channelManager) getReadyChannel(l1Head eth.BlockID) (*channel, error) {
...
-	// Short circuit if there is pending tx data or the channel manager is closed.
-	if dataPending || s.closed {
+	// Short circuit if there is pending tx data or the channel manager is closed
+	if dataPending {
 		return firstWithTxData, nil
 	}
 
+	if s.closed {
+		return nil, io.EOF
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: Why this new check? Why not at the top of the function?

File: op-batcher/batcher/channel_manager.go [hunk 12]
```diff
@@ -146,12 +157,53 @@ func (s *channelManager) nextTxData(channel *channel) (txData, error) {
...
+		"useBlobsAfter", newCfg.UseBlobs)
+	s.Requeue(newCfg)
+	channel, err = s.getReadyChannel(l1Head)
+	if err != nil {
+		return emptyTxData, err
+	}
+	return s.nextTxData(channel)
+}
+
+// getReadyChannel returns the next channel ready to submit data, or an error.
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go:195]: we could also add a comment that it adds blocks from the block queue into the currently open channel.

File: op-batcher/batcher/channel_manager.go [hunk 13]
```diff
@@ -414,3 +475,26 @@ func (s *channelManager) Close() error {
 	}
 	return nil
 }
+
+// Requeue rebuilds the channel manager state by
+// rewinding blocks back from the channel queue, and setting the defaultCfg.
+func (s *channelManager) Requeue(newCfg ChannelConfig) {
+	newChannelQueue := []*channel{}
+	blocksToRequeueInChannelManager := []*types.Block{}
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager.go]: I find the `InChannelManager` suffix superfluous and making the variable name unnecessarily long. We are in the context of the channel manager here, so I suggest we shorten it.

File: op-batcher/batcher/channel_manager_test.go (hunk 7 above)
Comment (sebastianst) [reply op-batcher/batcher/channel_manager_test.go]: agree, not sure if we benefit from a random seed

File: op-batcher/batcher/channel_manager_test.go [hunk 14]
```diff
@@ -483,3 +484,174 @@ func TestChannelManager_ChannelCreation(t *testing.T) {
...
+			// by possibly flipping the state of the
+			// fake channel provider
+			l.Info("updating market conditions", "chooseBlobs", tc.chooseBlobsWhenChannelSubmitted)
+			cfg.chooseBlobs = tc.chooseBlobsWhenChannelSubmitted
+
+			// Add a block and call TxData until
+			// we get some data to submit
+			var data txData
+			for {
+				m.blocks = []*types.Block{blockA}
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager_test.go:583]: You're using Brotli compression and always add the same block, so this compresses probably very well and can take a lot of iterations to fill up. Maybe better generate a new random block every iteration to fill up the channel more quickly?

**Review by sebastianst — COMMENTED** [review]
thanks for adding the additional tests!

File: op-batcher/batcher/channel_manager_test.go (hunk 14 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager_test.go:583]: Good point. I the end I went with the NoneCompressor [2a9b7ef](https://github.com/ethereum-optimism/optimism/pull/12002/commits/2a9b7efd1d58d3d70a255c889e8b3552df59239f) and also increased the size of the (same) block to fill the channel more quickly (after 3 blocks) [7d63970](https://github.com/ethereum-optimism/optimism/pull/12002/commits/7d639707e7632deebf8d4f37bd05366bae1567bb)

File: op-batcher/batcher/channel_manager_test.go (hunk 7 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager_test.go]: fixed in [2a9b7ef](https://github.com/ethereum-optimism/optimism/pull/12002/commits/2a9b7efd1d58d3d70a255c889e8b3552df59239f)

File: op-batcher/batcher/channel_manager.go (hunk 10 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: I agree, removed in [7b740f7](https://github.com/ethereum-optimism/optimism/pull/12002/commits/7b740f739d9b8bebd9e18d491184e231dadf5772)

File: op-batcher/batcher/channel_manager.go (hunk 12 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go:195]: [8d585cc](https://github.com/ethereum-optimism/optimism/pull/12002/commits/8d585cc35e7909eeff2d738c382c90f5f69881b1)

File: op-batcher/batcher/channel_manager.go (hunk 13 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: [a9b9e93](https://github.com/ethereum-optimism/optimism/pull/12002/commits/a9b9e93139866cc9018309ac788aef1cd6904820)

File: op-batcher/batcher/channel_manager.go (hunk 11 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: The existing behavior is such that it is possible to return `nil,nil` (when the channel manager is closed and when there is no channel found in the search for `firstWithTxData`). 

It is important to not return early unconditionally when the channel manager is closed, in case we are trying to still finish submitting a channel.

File: op-batcher/batcher/channel_manager.go (hunk 8 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: The motivation was to be able to differentiate the two different cases which return `io.EOF` to make testing easier. But it is not strictly necessary for this PR [510464d](https://github.com/ethereum-optimism/optimism/pull/12002/commits/510464d28cda1e5f0c2690be810ddda5c24a68c0)

Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: The driver does behave differently (logging behavior) with the change, as the error bubbles up: 

https://github.com/ethereum-optimism/optimism/blob/510464d28cda1e5f0c2690be810ddda5c24a68c0/op-batcher/batcher/driver.go#L536-L547

File: op-batcher/batcher/channel_manager.go (hunk 9 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: Reverted to the old behaviour [510464d](https://github.com/ethereum-optimism/optimism/pull/12002/commits/510464d28cda1e5f0c2690be810ddda5c24a68c0)

File: op-batcher/batcher/channel_manager.go (hunk 8 above)
Comment (sebastianst) [reply op-batcher/batcher/channel_manager.go]: The driver deliberately checks `io.EOF` here and only trace- instead of error-logs, because it's an expected sentinel error in the normal path of execution. So error logging on `ErrInsufficientData` would be wrong. Also wondering why, from the perspective of the caller of `TxData`, would it ever behave differently depending on `io.EOF` or `ErrInsufficientData`? It shouldn't care whether there's no data because there's no channel or because the current channel doesn't yield a frame yet.

File: op-batcher/batcher/channel_manager_test.go (hunk 14 above)
Comment (sebastianst) [reply op-batcher/batcher/channel_manager_test.go:583]: Even better! The NoneCompressor is a good choice in tests.

**Review by sebastianst — APPROVED** [review]

File: op-batcher/batcher/channel_manager.go (hunk 8 above)
Comment (geoknee) [reply op-batcher/batcher/channel_manager.go]: Agreed!
